src/esperanza/walletextension.cpp

```diff
@@ -561,16 +577,6 @@ void WalletExtension::VoteIfNeeded(const FinalizationState &state, const blockch
     return;
   }
 
-  const uint32_t dynasty = state.GetCurrentDynasty();
-
-  if (dynasty > validator->m_end_dynasty) {
-    return;
-  }
-
-  if (dynasty < validator->m_start_dynasty) {
-    return;
-  }
```

Comment:
Can you explain why these checks were removed?

Reply:
I guess I got it. Checks are done one level higher:
```cpp
  if (GetFinalizerPhase(*fin_state) == +ValidatorState::Phase::IS_VALIDATING) {
    VoteIfNeeded(*fin_state, tip_block_index->nHeight);
  }
```

Reply:
That's correct. Another explanation is to give clear responsibility to each function.
In `BlockConnected` we decide if finalizer should vote.
In `VoteIfNeeded` we vote unless this vote is duplicate/double/surrounded vote.

Reply:
@frolosofsky `VoteIfNeeded` is a private function and supposed to be called only from the `BlockConnected`. Technically everything can be written in one function but we split on two historically.

Reply:
Privateness is not a reason for confusing. Any function can be reused, by definition, and should have appropriate naming.

Reply:
@frolosofsky could you, please, share what your proposal is?

I am to guess the following:
1. Move everything that we have in BlockConnected to `VoteIfNeeded` and maybe remove `BlockConnected`
2. only rename `VoteIfNeeded` (my proposal is `VoteIfNotOffended` then) but would like to see your

Reply:
I suggest either

1. Keep old checks as is, then we do not have any subject of conversation.
2. Rework BlockConnected in a way to retrieve finalization state and validator, do minimal checks like `if (!validator)` and `assert(fin_state)`. Rework VoteIfNeeded in a way to accept finalization state and validator (atm we call `GetValidator` twice), apply the rest of checks, and finally vote.

I do not really like an idea of renaming VoteIfNeeded to VoteIfNotOffended. Even in this PR is has several checks, not only offending ones.

Reply:
@frolosofsky old checks are not enough to keep (they catch only a small part of the rule). We need to use `if (GetFinalizerPhase(*fin_state) == +ValidatorState::Phase::IS_VALIDATING)` which is now in `BlockConnected`. I propose to then invoke this condition as the first line inside of `VoteIfNeeded`.

WDYS?

Reply:
> old checks are not enough to keep (they catch only a small part of the rule). We need to use if (GetFinalizerPhase(*fin_state) == +ValidatorState::Phase::IS_VALIDATING) which is now in BlockConnected. I propose to then invoke this condition as the first line inside of VoteIfNeeded.

It sounds like a second approach: 

> Rework BlockConnected in a way to retrieve finalization state and validator, do minimal checks like if (!validator) and assert(fin_state). Rework VoteIfNeeded in a way to accept finalization state and validator (atm we call GetValidator twice), apply the rest of checks, and finally vote.